pysindy/optimizers/miosr.py

```diff
@@ -0,0 +1,248 @@
+        self.alpha = alpha
+        self.initial_guess = initial_guess
+        self.regression_timeout = regression_timeout
+        self.verbose = verbose
+
+        self.model = None
+
+    def _make_model(self, X, y, k, warm_start=None):
+        m = gp.Model()
+        n, d = X.shape
+        _, r = y.shape
+
+        beta = m.addMVar(
+            r * d, lb=-gp.GRB.INFINITY, vtype=gp.GRB.CONTINUOUS, name="beta"
+        )
+        iszero = m.addMVar(r * d, vtype=gp.GRB.BINARY, name="iszero")
+
+        # Sparsity constraint
+        for i in range(r * d):
+            m.addSOS(gp.GRB.SOS_TYPE1, [beta[i], iszero[i]])
+        m.addConstr(iszero.sum() >= (r * d) - k, name="sparsity")
+
+        # Group sparsity constraints
+        if self.target_sparsity is not None and self.group_sparsity is not None:
+            for i in range(r):
+                dimension_sparsity = self.group_sparsity[i]
+                print(dimension_sparsity)
+                m.addConstr(
+                    iszero[i * d : (i + 1) * d].sum() >= d - dimension_sparsity,
+                    name=f"group_sparsity{i}",
+                )
+
+        # General equality constraints
+        if self.constraint_lhs is not None and self.constraint_rhs is not None:
+            if self.constraint_order == "feature":
+                target_indexing = np.arange(r * d).reshape(r, d, order="F").flatten()
+                constraint_lhs = self.constraint_lhs[:, target_indexing]
+            else:
+                constraint_lhs = self.constraint_lhs
+            m.addConstr(
+                constraint_lhs @ beta == self.constraint_rhs, name="coefficient_constrs"
+            )
+
+        if warm_start is not None:
+            warm_start = warm_start.reshape(1, r * d)[0]
+            for i in range(d):
+                iszero[i].start = abs(warm_start[i]) < 1e-6
+                beta[i].start = warm_start[i]
+
+        Quad = np.dot(X.T, X)
+        obj = self.alpha * (beta @ beta)
+        for i in range(r):
+            lin = np.dot(y[:, i].T, X)
+            obj += beta[d * i : d * (i + 1)] @ Quad @ beta[d * i : d * (i + 1)]
+            obj -= 2 * (lin @ beta[d * i : d * (i + 1)])
+
+        m.setObjective(obj, gp.GRB.MINIMIZE)
+
+        m.params.OutputFlag = 1 if self.verbose else 0
+        m.params.timelimit = self.regression_timeout
+        m.update()
+
+        self.model = m
+
+        return m, beta
+
+    def _regress(self, X, y, k, warm_start=None):
+        """
+        Deploy and optimize the MIO formulation of L0-Regression.
+        """
+        m, beta = self._make_model(X, y, k, warm_start)
+        m.optimize()
+        return beta.X
+
+    def _reduce(self, x, y):
+        """
+        Runs MIOSR either per dimension or jointly on all dimensions.
+
+        Assumes an initial guess for coefficients and support are saved in
+        ``self.coef_`` and ``self.ind_``.
+        """
+        regress_jointly = self.target_sparsity is not None
```

Comment:
Nit: These kind of single-use boolean flags have a tendency to migrate further away from where they matter as later changes add lines of code.  Just inline it as `if target_sparsity is not None:` and add `# regress_jointly` as a comment.  (if that is, in fact, the desired behavior... see comment on the class docstring).